adapters/improvedigital/improvedigital.go

```diff
@@ -180,31 +184,73 @@ func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server co
 	return bidder, nil
 }
 
-func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
-	for _, imp := range imps {
-		if imp.ID == impID {
-			if imp.Banner != nil {
-				return openrtb_ext.BidTypeBanner, nil
-			}
-
-			if imp.Video != nil {
-				return openrtb_ext.BidTypeVideo, nil
-			}
+func getBidType(bid openrtb2.Bid, impMap map[string]openrtb2.Imp) (openrtb_ext.BidType, error) {
+	// there must be a matching imp against bid.ImpID
+	imp, found := impMap[bid.ImpID]
+	if !found {
+		return "", &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Failed to find impression for ID: \"%s\"", bid.ImpID),
+		}
+	}
 
-			if imp.Native != nil {
-				return openrtb_ext.BidTypeNative, nil
+	// if MType is not set in server response, try to determine it
+	if bid.MType == 0 {
+		if !isMultiFormatImp(imp) {
+			// Not a bid for multi format impression. So, determine MType from impression
+			if imp.Banner != nil {
+				bid.MType = openrtb2.MarkupBanner
+			} else if imp.Video != nil {
+				bid.MType = openrtb2.MarkupVideo
+			} else if imp.Audio != nil {
+				bid.MType = openrtb2.MarkupAudio
+			} else if imp.Native != nil {
+				bid.MType = openrtb2.MarkupNative
+			} else { // This should not happen.
+				// Let's handle it just in case by returning an error.
+				return "", &errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Could not determine MType from impression with ID: \"%s\"", bid.ImpID),
+				}
```

Comment:
As per https://github.com/prebid/prebid-server/pull/3077#issuecomment-1732901588 this error path is not covered. Should add supplemental tests for error path coverage

Reply:
Added supplemental test as suggested, in `adapters/improvedigital/improvedigitaltest/supplemental/missing_and_unsupported_mtype.json`